### PR TITLE
Add `File#chown` and `#chmod`

### PR DESCRIPTION
**Pull request by didactic-drunk**
Described in #9192

### CODE REVIEW

File: src/crystal/system/win32/file.cr [hunk 1]
```diff
@@ -135,6 +135,10 @@ module Crystal::System::File
     raise NotImplementedError.new("File.chown")
   end
 
+  def self.fchown(path : String, fd : Int, uid : Int32, gid : Int32) : Nil
+    raise NotImplementedError.new("File##chown")
```
Comment (straight-shoota) [inline src/crystal/system/win32/file.cr:139]:
```suggestion
    raise NotImplementedError.new("File#chown")
```

**Review by straight-shoota — COMMENTED** [review]
Could you please add cross-references between the docs for the new instance methods and the respective class methods?

**Comment by didactic-drunk** [comment]
> Could you please add cross-references between the docs for the new instance methods and the respective class methods?

I'm not sure of the format you want. `File.#info` & `File.#delete` don't have cross references. Do you have an example?

**Comment by straight-shoota** [comment]
Something like this maybe: [`57bb5ae` (#11881)](https://github.com/crystal-lang/crystal/pull/11881/commits/57bb5ae2a49585618dba2933becb3da2f8d383c8)

**Comment by didactic-drunk** [comment]
MacOS broke, not my fault?

```
 error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
```

**Comment by didactic-drunk** [comment]
1.5? Bump

**Review by straight-shoota — APPROVED** [review]

**Comment by didactic-drunk** [comment]
I'm not sure how to do syscall feature detection in the last commit.  Most systems support `futimens` except MacOS which only has `futimes`.

File: src/crystal/system/unix/file.cr [hunk 2]
```diff
@@ -179,6 +190,33 @@ module Crystal::System::File
...
+            timespecs[1] = to_timespec(mtime)
+            LibC.futimens(fd, timespecs)
+          elsif LibC.responds_to?(:futimes)
+            timevals = uninitialized LibC::Timeval[2]
+            timevals[0] = to_timeval(atime)
+            timevals[1] = to_timeval(mtime)
+            LibC.futimes(fd, timevals)
+          else
+            raise NotImplementedError.new("Missing futimens & futimes")
+          end
```
Comment (Sija) [inline src/crystal/system/unix/file.cr:206]:
```suggestion
    ret = {% if LibC.has_method?("futimens") %}
            timespecs = uninitialized LibC::Timespec[2]
            timespecs[0] = to_timespec(atime)
            timespecs[1] = to_timespec(mtime)
            LibC.futimens(fd, timespecs)
          {% elsif LibC.has_method?("futimes") %}
            timevals = uninitialized LibC::Timeval[2]
            timevals[0] = to_timeval(atime)
            timevals[1] = to_timeval(mtime)
            LibC.futimes(fd, timevals)
          {% else %}
            raise NotImplementedError.new("Missing futimens & futimes")
          {% end %}
```

**Review by beta-ziliani — COMMENTED** [review]
I only have one question, and if CI is ok, then nothing else :ok_hand:

File: src/file.cr [hunk 3]
```diff
@@ -731,10 +739,11 @@ class File < IO::FileDescriptor
       open(dst, "wb") do |d|
         # TODO use sendfile or copy_file_range syscall. See #8926, #8919
         IO.copy(s, d)
-      end
+        d.flush
```
[the remark this answers is not in the document]
Comment (didactic-drunk) [reply src/file.cr:754]: The fear was how windows works if permissions are changed to readonly (when copying a readonly file) before the file is fully written.  Idk if it's necessary.
 
POSIX doesn't need it.

[the remark this answers is not in the document]
Comment (beta-ziliani) [reply src/file.cr:754]: Sounds like the wise thing to do :tophat: 

https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-flushfilebuffers
> The file handle must have the GENERIC_WRITE access right.

OK, WASI is failing. I can fix that if you prefer.

[the remark this answers is not in the document]
Comment (didactic-drunk) [reply src/file.cr:754]: Please

**Review by beta-ziliani — APPROVED** [review]
:rocket:

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**